text/00110-compatibility-strategy.md

```diff
@@ -0,0 +1,364 @@
+
+<!-- To edit: URL Decode -> Edit -> URL Encode -->
+
+<img src='https://g.gravizo.com/svg?digraph%20G%20%7B%0A%20%20%20%20aize%20%3D%224%2C4%22%3B%0A%20%20%20%20%22merge-to-master%22%3B%0A%20%20%20%20%22merge-to-master%22%20-%3E%20%22test-framework-compatibility.sh%22%3B%0A%20%20%20%20node%20%5Bshape%3Dbox%2Cstyle%3Dfilled%2Ccolor%3D%22yellowgreen%22%5D%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20pass%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22pass-on-error%22%3B%0A%20%20%20%20node%20%5Bshape%3Dbox%2Cstyle%3Dfilled%2Ccolor%3D%22crimson%22%5D%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22fail%22%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22fail-on-error%22%3B%0A%20%20%20%20%22fail%22%20-%3E%20%22merge-to-master%22%20%5Blabel%20%3D%20%22%20%20make-compatible%5CnOR%5Cnmark-incompatible%22%5D%0A%20%20%20%20%22fail-on-error%22%20-%3E%20%22merge-to-master%22%20%5Blabel%20%3D%20%22%20%20fix-version-bump%22%5D%0A%20%20%7D%0A'/>
+
+<br>
+
+# Drawbacks
+
+- Dramatically increases the duration of our integration tests suite.
+- Requires manually editing the framework compatibility test file.
+
+    > This isn't necessarily a drawback, as it provides a clear paper trail of all our breakages.
+
+# Rationale and Alternatives
+
+I'm not seeing any substantially different design approaches. Not matter what, to implement these validations, we have to implement the `run_integration_tests_perm` function, which is the bulk of the work. Any other approach would probably result in an incomplete solution. 
+
+This design provides a robust mechanism that should serve us well for the forseeable future. It covers all our compatibility promises without the need to add specific (functionality) tests.
+
+# Adoption Strategy
+
+Adoption of this strategy should be rather seamless for developers. Like we already mentioned, the changes a developer would need to incorporate, will be guided by failing tests. Hopefully, the process will not require any memorization (or even documentation). 
+
+Having said that, this RFC can be a good reference documentation that provides rational for every step in the process.
+
+# Unresolved questions
+
+There are some scenarios which are still a bit vague with regards to what should be our expected outcome.
+
+## Accidental (In)Compatibility
+
+Imagine a developer pushes some code that changes the `manifest.json`, without bumping the protocol version. Also imagine that this change is compatible with older CLI versions (maybe just an added property), meaning that all integration tests pass. Actually, you don't have to imagine, here is a real life example: [custom docker files support](https://github.com/aws/aws-cdk/pull/5652). Note that this feature only requires a new CLI to support a **new** feature, while all previous functionality remains intact.
+
+With this proposed mechanism, its very likely this code will be released without performing a version bump.
```

Comment:
I believe this is a use case we should somehow "catch". I would have expected the new integration test to somehow fail when tests would be executed with the old CLI...

Reply:
I agree - though it wasn't the main focus of this RFC. Off the top of my head:

- If we abandon cx-protocol versioning altogether, this problem is solved.
- If we stick with it, we can run **latest** integration tests, with a **new** framework version and an **old** cli. In the given example, this suite will fail because the cli won't use the correct Dockerfile (which is what the new test we add validates).